src/org/sosy_lab/common/NativeLibraries.java

```diff
@@ -105,55 +111,62 @@ public static Architecture guessVmArchitecture() {
-        prop = System.getProperty("sun.arch.data.model");
+      currentArch = downgradeArchIfRequired(getArchFromSystemProperty());
+
+      return currentArch;
+    }
+
+    private static Architecture getArchFromSystemProperty() {
+      @Var String osArch = StandardSystemProperty.OS_ARCH.value();
+      if (isNullOrEmpty(osArch)) {
+        throw new UnsatisfiedLinkError("No value for os.arch" + REPORT_MESSAGE);
       }
 
-      if (!isNullOrEmpty(prop)) {
-        switch (prop) {
-          case "32":
-            currentArch = Architecture.X86;
-            break;
-          case "64":
-            currentArch = Architecture.X86_64;
-            break;
-          default:
-            throw new UnsatisfiedLinkError(
-                "Unknown value for os.arch.data.model: '"
-                    + prop
-                    + "', please report this together with information about your system "
-                    + "(OS, architecture, JVM).");
-        }
-      } else {
+      osArch = Ascii.toLowerCase(osArch.replace(" ", ""));
+
+      switch (osArch) {
+        case "i386":
+        case "i686":
+        case "x86":
+          return Architecture.X86;
+        case "amd64":
+        case "x86_64":
+          return Architecture.X86_64;
+        case "arm":
+        case "armv6":
+        case "armv7":
+          return Architecture.ARM;
+        case "aarch64":
+          return Architecture.ARM64;
+        default:
+          throw new UnsatisfiedLinkError(
+              "Unknown value for os.arch: '"
+                  + StandardSystemProperty.OS_ARCH.value()
+                  + "'"
+                  + REPORT_MESSAGE);
+      }
+    }
 
-        prop = StandardSystemProperty.JAVA_VM_NAME.value();
-        if (!isNullOrEmpty(prop)) {
-          prop = Ascii.toLowerCase(prop);
-
-          if (prop.contains("32-bit") || prop.contains("32bit") || prop.contains("i386")) {
-
-            currentArch = Architecture.X86;
-          } else if (prop.contains("64-bit")
-              || prop.contains("64bit")
-              || prop.contains("x64")
-              || prop.contains("x86_64")
-              || prop.contains("amd64")) {
-
-            currentArch = Architecture.X86_64;
-          } else {
-            throw new UnsatisfiedLinkError(
-                "Unknown value for java.vm.name: '"
-                    + prop
-                    + "', please report this together with information about your system "
-                    + "(OS, architecture, JVM).");
-          }
-        } else {
-          throw new UnsatisfiedLinkError("Could not detect system architecture");
+    /**
+     * In most cases, the system architecture matches the JVM architecture. In some rare cases,
+     * architecture from 'os.arch' might not be sufficient. For example, we could use an emulator
+     * and execute a 32bit JVM on a 64bit system. We cannot check each possible combination and
+     * emulator, but check only the important ones.
```

Comment:
Hm, I think there is a misconception here. This comment seems to indicate that running a JVM that does not have exactly the same architecture as the OS is a special edge case that involves emulators.

But this is not the case! Running for example a 32-bit JVM on an 64-bit OS is a totally common case that is supported out-of-the box, no special setup or emulators involved.

What the code behind `guessVmArchitecture()` needs to do is to find out the *JVM* architecture, the OS architecture is not really relevant. The current code seems look at the latter, and then tries some adjustments for the JVM architecture. Can we do better, i.e., look at the JVM architecture primarily?

Reply:
The text contains the wording "for example ... emulator". Furthermore, executing 32bit application on a 64bit CPU requires the CPU to "redirect/simulate/emulate" the commands and registers, which is typically done quite performant directly in hardware.

Most developers install a JDK in 32 or 64bit according to their CPU architecture. Java in 32bit is outdated, as even embedded devices are available with more than 4 GB of memory, which is the maximum addressable space for a 32bit application. And the other way, from 32 to 64bit, does not work anyway.

Overall, I do not see a negative impact of my code changes for default machines and guessing architecture. For the ARM-part, we can discuss changes, see below.

Reply:
My point was that the comment gives a misleading representation of the importance of these code paths, which could easily lead to future developers ignoring this and introducing errors. Using a 32-bit JVM is sometimes done on purpose for saving memory if the application would not need more than a few GB anyway, so it is not like that 32-bit is exclusively a thing from the past.

The current comment is better.